File: FBAudio-Android/app/src/main/java/com/fba/app/data/local/AppSettings.kt
package com.fba.app.data.local

import android.content.Context
import android.content.SharedPreferences
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow

/** User settings shown on My FBA. Backed by SharedPreferences, exposed as flows. */
class AppSettings(context: Context) {
    private val prefs: SharedPreferences = context.getSharedPreferences("app_settings", Context.MODE_PRIVATE)

    private val _englishOnly = MutableStateFlow(prefs.getBoolean(KEY_ENGLISH_ONLY, true))
    /** Hide talks and lists in languages other than English (default on). */
    val englishOnly: StateFlow<Boolean> = _englishOnly

    private val _preferRemastered = MutableStateFlow(prefs.getBoolean(KEY_PREFER_REMASTERED, true))
    /** Play the remastered version when a talk has one (default on). */
    val preferRemastered: StateFlow<Boolean> = _preferRemastered

    fun setEnglishOnly(value: Boolean) {
        prefs.edit().putBoolean(KEY_ENGLISH_ONLY, value).apply()
        _englishOnly.value = value
    }

    fun setPreferRemastered(value: Boolean) {
        prefs.edit().putBoolean(KEY_PREFER_REMASTERED, value).apply()
        _preferRemastered.value = value
    }

    /** Per-talk override of the remastered/original choice; null = follow the setting. */
    fun remasterChoice(catNum: String): Boolean? =
        if (prefs.contains("remaster_$catNum")) prefs.getBoolean("remaster_$catNum", true) else null

    fun setRemasterChoice(catNum: String, useRemaster: Boolean) {
        prefs.edit().putBoolean("remaster_$catNum", useRemaster).apply()
    }

    /** Resolved choice for a talk: explicit per-talk choice, else the global default. */
    fun useRemaster(catNum: String): Boolean = remasterChoice(catNum) ?: preferRemastered.value

    companion object {
        private const val KEY_ENGLISH_ONLY = "english_only"
        private const val KEY_PREFER_REMASTERED = "prefer_remastered"
    }
}

File: FBAudio-iOS/FBAudio/Data/AppSettings.swift
import Foundation
import Combine

/// User settings shown on My FBA. Backed by UserDefaults.
@MainActor
final class AppSettings: ObservableObject {
    static let shared = AppSettings()

    private let defaults = UserDefaults.standard
    private static let englishOnlyKey = "english_only"
    private static let preferRemasteredKey = "prefer_remastered"

    /// Hide talks and lists in languages other than English (default on).
    @Published var englishOnly: Bool {
        didSet { defaults.set(englishOnly, forKey: Self.englishOnlyKey) }
    }

    /// Play the remastered version when a talk has one (default on).
    @Published var preferRemastered: Bool {
        didSet { defaults.set(preferRemastered, forKey: Self.preferRemasteredKey) }
    }

    init() {
        englishOnly = defaults.object(forKey: Self.englishOnlyKey) as? Bool ?? true
        preferRemastered = defaults.object(forKey: Self.preferRemasteredKey) as? Bool ?? true
    }

    /// Per-talk override of the remastered/original choice; nil = follow the setting.
    func remasterChoice(_ catNum: String) -> Bool? {
        defaults.object(forKey: "remaster_\(catNum)") as? Bool
    }

    func setRemasterChoice(_ catNum: String, useRemaster: Bool) {
        defaults.set(useRemaster, forKey: "remaster_\(catNum)")
    }

    /// Resolved choice for a talk: explicit per-talk choice, else the global default.
    func useRemaster(_ catNum: String) -> Bool {
        remasterChoice(catNum) ?? preferRemastered
    }
}
